### PR TITLE
Python3: Changes needed to run runTheMatrix with python2/3

### PR DESCRIPTION
**Pull request by smuzaffar**
#### PR description:

Changes needed to run runTheMatrix with python3

#### PR validation:

Locally runing of runTheMatix (to dump relval commands) was successful. Order of command-line arguments to cmsDriver.py is not identical (because of change of `dict` behaviour in python3)

- **python2 runTheMatrix.py -n -e -l 137.8 -i all**
```
found a total of  794  workflows:
      of which the following 1 were selected:

137.8  RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+RunEGamma2018D+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM_MULTIRUN [1]: input from: /EGamma/Run2018C-v1/RAW with run [] 
                                           [2]: cmsDriver.py step2  --datatier FEVTDEBUGHLT --conditions auto:run2_hlt_relval -s L1REPACK:Full,HLT:@relval2018 --process reHLT --data  --era Run2_2018 --eventcontent FEVTDEBUGHLT -n 100 
                                           [3]: cmsDriver.py step3  --conditions auto:run2_data -s RAW2DIGI,L1Reco,RECO,SKIM:ZElectron+SinglePhotonJetPlusHOFilter+EXOMONOPOLE,EI,PAT,ALCA:SiStripCalZeroBias+SiStripCalMinBias+SiStripCalSmallBiasScan+TkAlMinBias+EcalESAlign,DQM:@standardDQMFakeHLT+@miniAODDQM+@L1TEgamma --runUnscheduled  --process reRECO --data  --era Run2_2018 --eventcontent RECO,MINIAOD,DQM --hltProcess reHLT --scenario pp --datatier RECO,MINIAOD,DQMIO --customise Configuration/DataProcessing/RecoTLR.customisePostEra_Run2_2018 -n 100 
                                           [4]: input from: /EGamma/Run2018D-v1/RAW with run []
                                           [5]: cmsDriver.py step5  --datatier FEVTDEBUGHLT --conditions auto:run2_hlt_relval -s L1REPACK:Full,HLT:@relval2018 --process reHLT --data  --era Run2_2018 --eventcontent FEVTDEBUGHLT -n 100 
                                           [6]: cmsDriver.py step6  --conditions auto:run2_data_promptlike -s RAW2DIGI,L1Reco,RECO,SKIM:ZElectron+SinglePhotonJetPlusHOFilter+EXOMONOPOLE,EI,PAT,ALCA:SiStripCalZeroBias+SiStripCalMinBias+SiStripCalSmallBiasScan+TkAlMinBias+EcalESAlign,DQM:@standardDQMFakeHLT+@miniAODDQM+@L1TEgamma --runUnscheduled  --process reRECO --data  --era Run2_2018 --eventcontent RECO,MINIAOD,DQM --hltProcess reHLT --scenario pp --datatier RECO,MINIAOD,DQMIO --customise Configuration/DataProcessing/RecoTLR.customisePostEra_Run2_2018 -n 100 
                                           [7]: cmsDriver.py step7  --conditions auto:run2_data -s HARVESTING:@standardDQMFakeHLT+@miniAODDQM+@L1TEgamma --data  --era Run2_2018 --filein file:step6_inDQM.root,file:step3_inDQM.root --scenario pp --filetype DQM --customise Configuration/StandardSequences/DQMSaverAtJobEnd_cff -n 100 

1 workflows with 7 steps

```
- **python3 runTheMatrix.py -n -e -l 137.8 -i all**
```
found a total of  794  workflows:
      of which the following 1 were selected:

137.8  RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+RunEGamma2018D+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM_MULTIRUN [1]: input from: /EGamma/Run2018C-v1/RAW with run [] 
                                           [2]: cmsDriver.py step2  --process reHLT -s L1REPACK:Full,HLT:@relval2018 --conditions auto:run2_hlt_relval --data  --eventcontent FEVTDEBUGHLT --datatier FEVTDEBUGHLT --era Run2_2018 -n 100 
                                           [3]: cmsDriver.py step3  --runUnscheduled  --conditions auto:run2_data -s RAW2DIGI,L1Reco,RECO,SKIM:ZElectron+SinglePhotonJetPlusHOFilter+EXOMONOPOLE,EI,PAT,ALCA:SiStripCalZeroBias+SiStripCalMinBias+SiStripCalSmallBiasScan+TkAlMinBias+EcalESAlign,DQM:@standardDQMFakeHLT+@miniAODDQM+@L1TEgamma --datatier RECO,MINIAOD,DQMIO --eventcontent RECO,MINIAOD,DQM --data  --process reRECO --scenario pp --era Run2_2018 --customise Configuration/DataProcessing/RecoTLR.customisePostEra_Run2_2018 --hltProcess reHLT -n 100 
                                           [4]: input from: /EGamma/Run2018D-v1/RAW with run []
                                           [5]: cmsDriver.py step5  --process reHLT -s L1REPACK:Full,HLT:@relval2018 --conditions auto:run2_hlt_relval --data  --eventcontent FEVTDEBUGHLT --datatier FEVTDEBUGHLT --era Run2_2018 -n 100 
                                           [6]: cmsDriver.py step6  --runUnscheduled  --conditions auto:run2_data_promptlike -s RAW2DIGI,L1Reco,RECO,SKIM:ZElectron+SinglePhotonJetPlusHOFilter+EXOMONOPOLE,EI,PAT,ALCA:SiStripCalZeroBias+SiStripCalMinBias+SiStripCalSmallBiasScan+TkAlMinBias+EcalESAlign,DQM:@standardDQMFakeHLT+@miniAODDQM+@L1TEgamma --datatier RECO,MINIAOD,DQMIO --eventcontent RECO,MINIAOD,DQM --data  --process reRECO --scenario pp --era Run2_2018 --customise Configuration/DataProcessing/RecoTLR.customisePostEra_Run2_2018 --hltProcess reHLT -n 100 
                                           [7]: cmsDriver.py step7  -s HARVESTING:@standardDQMFakeHLT+@miniAODDQM+@L1TEgamma --conditions auto:run2_data --data  --filetype DQM --scenario pp --era Run2_2018 --customise Configuration/StandardSequences/DQMSaverAtJobEnd_cff --filein file:step6_inDQM.root,file:step3_inDQM.root -n 100 

1 workflows with 7 steps
```

### CODE REVIEW

**Comment by smuzaffar** [comment]
FYI @davidlange6

**Comment by smuzaffar** [comment]
please test

File: Configuration/PyReleaseValidation/scripts/das-selected-lumis.py [hunk 1]
```diff
@@ -16,8 +16,8 @@
 lumi_data = lumi_data['data']
 
 def match_in(sub_list,lumi_list):
-  sub_list = map(int,sub_list)
-  lumi_list = map(int,lumi_list)
+  sub_list = list(map(int,sub_list))
```
Comment (davidlange6) [inline Configuration/PyReleaseValidation/scripts/das-selected-lumis.py]: this starts to be a complicated way of writing
for i in range(int(sub_list[0]),int(sub_list[1])+1):
   if i >= int(lumi_list[0]) and i <= int(lumi_list[1]): return True
return False

**Comment by davidlange6** [comment]
thanks @smuzaffar

File: Configuration/PyReleaseValidation/scripts/das-selected-lumis.py (hunk 1 above)
Comment (smuzaffar) [reply Configuration/PyReleaseValidation/scripts/das-selected-lumis.py]: thanks for suggestion. I have updated it

**Comment by smuzaffar** [comment]
please test

File: Configuration/PyReleaseValidation/python/relval_steps.py [hunk 2]
```diff
@@ -3047,9 +3047,9 @@ def gen2018HiMix(fragment,howMuch):
 defaultDataSets['2026D35']='CMSSW_10_4_0_mtd3-103X_upgrade2023_realistic_v2_2023D35noPU_2-v'
 defaultDataSets['2026D41']='CMSSW_10_6_0_patch2-106X_upgrade2023_realistic_v3_2023D41noPU-v'
 
-keys=defaultDataSets.keys()
-for key in keys:
-  defaultDataSets[key+'PU']=defaultDataSets[key]
+puDataSets = {}
```
Comment (kpedro88) [inline Configuration/PyReleaseValidation/python/relval_steps.py]: for my own understanding, what in python3 makes this change necessary?

**Comment by fabiocos** [comment]
@smuzaffar I do not see the order of output options as a relevant issue

File: Configuration/PyReleaseValidation/python/relval_steps.py (hunk 2 above)
Comment (smuzaffar) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: python3 complain about change in dict size 
```
RuntimeError: dictionary changed size during iteration
```

Comment (kpedro88) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: Ah, so it must be a change in how `.keys()` behaves. Good to know, thanks.

**Comment by kpedro88** [comment]
+upgrade

**Comment by davidlange6** [comment]
yes, it returns an iterator rather than a list in python3

**Comment by smuzaffar** [comment]
@davidlange6, I think we can just do `keys=list(defaultDataSets.keys())` to make it work for both python2/3. Would you like me to update it? I have to update this PR any way to fix some more tools

**Comment by davidlange6** [comment]
i had avoided that since it slows down python2. I think your current implementation is good.

**Comment by smuzaffar** [comment]
ok then I will not update this PR and will make a new PR to fix
```
CondFormats/Serialization/python/condformats_serialization_generate.py
FWCore/Utilities/scripts/edmCheckClassTransients
FWCore/Utilities/scripts/edmCheckClassVersion
```

**Comment by fabiocos** [comment]
discussing with @smuzaffar I want to record an extra comment: the order of options is not relevant unless there are duplicated ones, and this normally happens when some customisation commands are used. The most trivial case I have in mind is changing the number of events to be processed. In that acse the order matter, and we need to ensure it stays the desired one

**Comment by davidlange6** [comment]
Code whose functionality depends on the order of iterating through a dict is buggy. Command line options are instead parsed from a string that defines the order in which things are parsed. But maybe you have a specific usage to be concerned about?

Cheers,
David


On 20 Sep 2019, at 18:23, Fabio Cossutti <notifications@github.com<mailto:notifications@github.com>> wrote:


discussing with @smuzaffar<https://github.com/smuzaffar> I want to record an extra comment: the order of options is not relevant unless there are duplicated ones, and this normally happens when some customisation commands are used. The most trivial case I have in mind is changing the number of events to be processed. In that acse the order matter, and we need to ensure it stays the desired one

**Comment by fabiocos** [comment]
@davidlange6 I was just referring to the comment of @smuzaffar about the change of option order in the built command. The use case I have trivially in mind is 
```
runTheMatrix.py -l 503 --command "-n 100" 
```
which moves the number of events of the workflow from the default value (10) to 100. I have verified that the order is not changed in this case. BTW, I see that for the option parsing we are using optparse which, according to https://docs.python.org/2/library/optparse.html is already deprecated in python 2.7 in favour of argparse.

**Comment by fabiocos** [comment]
@smuzaffar as far as I can see there in my test is no issue even with python3 when overriding the command

**Comment by smuzaffar** [comment]
@fabiocos , yes I also have run tests with `--command '-n 1'` and both python2 and python3 keep `-n 1` at the end.

**Comment by smuzaffar** [comment]
+core
@fabiocos , can we get this in.

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
+operations

the changes in Configuration/Applications look correct

**Comment by chayanit** [comment]
+1

**Comment by smuzaffar** [comment]
hold
we need https://github.com/cms-sw/cmsdist/pull/5243 to go in first.

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
I have moved back to old commit so that we can integrate this PR. I will create a separate PR  to fix some more python3 related errors.

**Comment by smuzaffar** [comment]
unhold

**Comment by fabiocos** [comment]
+1

it looks ready to be integrated, and with this a py3-based IB may be prepared for further tests

**Pull request merged**